### ISSUE TITLE
Shader instance parameters conflict with each other when using Next Pass

### ISSUE THREAD

**Issue by guessy42** [godotengine/godot#78945]
### Godot version

v4.0.3.stable.mono.official [5222a99f5]

### System information

Windows 10, Vulkan (Forward+), NVIDIA GeForce GTX 1050 Ti (driver verison 516.94)

### Issue description

If you have two shaders A and B, each with their own instance parameters, when one is setup as the next pass of the other, their instance parameters will display correctly in the editor but conflict when actually rendering; going by index, whichever parameter is last edited becomes the one used visually in both materials.

This is an example testcase where the next pass is a simple outline shader:
![Godot_v4 0 3-stable_mono_win64_nWdhh4fTYc](https://github.com/godotengine/godot/assets/121681708/97575da5-3d23-41ba-a6f8-24e97179020a)
As you might note, the various instance parameters are all correctly displayed in the inspector, but only two are being used visually at any one time. This persists through saving and reloading or restarting the editor.

Changing the first instance parameter of one shader will change the first instance parameter for the both of them, ditto the second and so on.

### Steps to reproduce

Simply try changing the instance parameters in the example project and you'll see the issue.

### Minimal reproduction project

[ShaderInstanceParamBug.zip](https://github.com/godotengine/godot/files/11929333/ShaderInstanceParamBug.zip)

**Comment by AThousandShips** [comment]
This sounds like the limitations that are listed [here](https://docs.godotengine.org/en/stable/tutorials/shaders/shader_reference/shading_language.html#per-instance-uniforms)

**Comment by guessy42** [comment]
> This sounds like the limitations that are listed [here](https://docs.godotengine.org/en/stable/tutorials/shaders/shader_reference/shading_language.html#per-instance-uniforms)

I was under the impression that was related to unique parameter names? (eg a multi-material setup with two instance parameters sharing the same name are forced to share the same value). I don't believe I've ever noticed similar behaviour to this with a multi-material setup, but I can test for it.

**Comment by AThousandShips** [comment]
I'm not sure, does it happen if you assign specific indices to them?

**Comment by guessy42** [comment]
> I'm not sure, does it happen if you assign specific indices to them?

Oh rats, I never spotted that. Yes, assigning indices fixes things.

> I suspect this, and some bug relating to how the editor handles these:
> 
> > Per-instance uniforms are set on each GeometryInstance3D, rather than on each Material instance. Take this into account when working with meshes that have multiple materials assigned to them, or MultiMesh setups.

Just did a quick check and yeah, multi-material setups actually do display the exact same behaviour as the Next Pass bug. Here's the updated MRP:
[ShaderInstanceParamBug2.zip](https://github.com/godotengine/godot/files/11929608/ShaderInstanceParamBug2.zip)

I feel like there should at least be a warning or something, yeah.

**Comment by AThousandShips** [comment]
I think you misread the documentation:
> If your mesh uses multiple materials, the parameters for the first mesh material found will "win" over the subsequent ones, unless they have the same name, index and type. In this case, all parameters are affected correctly.

I.e. it's not that they have to be identical if they share names, but they they will be ignored unless they have identical status

**Comment by guessy42** [comment]
What's weird is that the 'first material found' actually *doesn't* win in editor, it's whichever you edit last.

I'd argue the documentation is fine, it's the editor itself that needs to warn you if you have overlapping instance parameter indices. (I honestly don't ever remember there even being a part of the docs that talks about instance parameter indices, so I suspect I was going off of memory of an older version of the documentation)

**Comment by clayjohn** [comment]
I agree some form of warning would be preferable. Given how the system works, I'm not sure a warning is possible, but it is certainly worth investigating for whoever is interested. Ultimately this is a documentation issue. The problem is well described by the current docs, but it should likely be flagged in more places so that users have an easier time finding it

https://docs.godotengine.org/en/stable/tutorials/shaders/shader_reference/shading_language.html#per-instance-uniforms

**Comment by QbieShay** [comment]
Maybe a warning when there's multiple materials on the same nodes with different instance parameters?

**Comment by ghost** [comment]
> Maybe a warning when there's multiple materials on the same nodes with different instance parameters?

Yea, that'll work. I'll try to open a PR for this today.
